### ISSUE TITLE
how to convert a pre-trained model (frozen graph, savedmodel or ckpt) to a new tf-trt-optimized savedmodel for tensorflow:serving

### ISSUE THREAD

**Issue by Eloring** [tensorflow/tensorrt#86]
**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux RHEL 7.6
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below):1.13.1
- Python version: Python 2.7.16 :: Anaconda, Inc.
- CUDA/cuDNN version: CUDA 10.1, cuDNN 7.6
- GPU model and memory: 2 Tesla K80, 12G/12G

**Describe the current behavior**

Use the tool saved_model_cli to covert a severable model with TF-TRT:
~~~
saved_model_cli convert \
--dir "/path/to/mask_rcnn/saved_model" \
--output_dir "/path/to/trt-mask-rcnn" \
--tag_set serve \
tensorrt --precision_mode FP32 --max_batch_size 32 --is_dynamic_op True
~~~

## While I got a .pb graph and the variables fold is empty, which can't be launched on tensorflow:serving.
~~~
variables/
    NULL
saved_model.pb
~~~

**Describe the expected behavior**

The following structure is expected. https://www.tensorflow.org/guide/saved_model#structure_of_a_savedmodel_directory

~~~
variables/
    variables.data-?????-of-?????
    variables.index
saved_model.pb|saved_model.pbtxt
~~~

My final target is to [Optimizing TensorFlow Serving performance with NVIDIA TensorRT](https://medium.com/tensorflow/optimizing-tensorflow-serving-performance-with-nvidia-tensorrt-6d8a2347869a), and the model that I want to launch is a pre-trained mask-rcnn downloaded from [tensorflow_model_zoo](https://github.com/tensorflow/models/blob/master/research/object_detection/g3doc/detection_model_zoo.md)

I have successfully launched several models without TensorRT, while how can I accelerate the inference with the help of TensorRT.

Any help would be grateful!

**Comment by xxllp** [comment]
the  saved model you  get can use by tf -serving  ,you can try it out

**Comment by Eloring** [comment]
@xxllp Thanks!
Although the variables fold is empty, the converted model can be launched on tensorflow:serving.
The procedures are:
    1. prepare a pre-trained model (.ckpt and .config files)
    2. export to a saved_model (.pb and variables), which can be launched on tf:serving
    3. convert to another saved_model (only .pb) with TensorFlow-TensorRT
My questions is, in the 2nd step, only saved_model.pb can not be launched, but in the 3rd step, it can.
Why?

**Comment by xxllp** [comment]
看一下官方介绍才行哈，2 里面的两个文件 合并才有3的，参数固定化了

**Comment by Eloring** [comment]
@xxllp 谢谢！

**Comment by xxllp** [comment]
这个测试结果有啥提高吗

**Comment by Eloring** [comment]
不高。。
Platform |Model |  Img amount | Img Size | Batch size   | mAP | Latency [load image + inference + save result]
-- | -- |  -- |-- | -- | -- | --
TF:Serving|Faster RCNN | 4952  | 300*300  | 32  | 0.7021 | 419204 ms
TF:Serving +TRT| Faster RCNN | 4952  | 300*300  | 32  | 0.7019 | 402723 ms

**Comment by Eloring** [comment]
@xxllp 这是mask rcnn的结果，latency没有很大提高，其他的performance我还没测

Platform |Model |  Img amount | Img Size | Batch size   | mAP | Latency [load image + inference + save result]
-- | -- |  -- |-- | -- | -- | --
TF:Serving|Mask RCNN | 4952  | 300*300  | 32  | 0.6977 | 1085690 ms
TF:Serving +TRT| Mask RCNN | 4952  | 300*300  | 32  | 0.6974 | 1073212 ms

**Comment by MachineJeff** [comment]
你是直接 docker 把它pull下来做的，还是单独安装tensorrt之后做的？

**Comment by Eloring** [comment]
@MachineJeff 自己编译的tensorflow和tensorrt。我后来发现没有提高是因为机器不支持FP16和INT8，换了V100就好了（（哈哈哈哈

**Comment by MachineJeff** [comment]
@Eloring 啊哈哈哈 好的。顺便问下，自己编译tensorflow和tensorrt是指用源码安装么？ bazel？

**Comment by Eloring** [comment]
@MachineJeff 对。

**Comment by MachineJeff** [comment]
@Eloring OK！我也试一下，docker镜像真的是太慢了，没有国内的源

**Comment by MachineJeff** [comment]
@Eloring 等等，最后一个问题，TF是1.14么？哪个版本开始支持tensorrt

**Comment by Eloring** [comment]
1.14好用点～

**Comment by MachineJeff** [comment]
@Eloring 好滴~谢谢您

**Comment by MachineJeff** [comment]
@Eloring  哈喽 我用自己编译的环境 和 docker环境都做了自己模型的实验，发现并没有提速（T4的卡） #148 您能不能帮帮忙

**Comment by elvys-zhang** [comment]
@MachineJeff 如果没有提速的话，看下转出来的模型中有trt的node吗？
```
    with tf.Session() as sess:
        meta_graph_def = tf.saved_model.loader.load(
                sess,
                [tf.saved_model.tag_constants.SERVING],
                '/path/to/your/model')
        print([n.name for n in meta_graph_def.graph_def.node if 'TRTEngine' in n.name])
```

**Comment by Eloring** [comment]
@MachineJeff 我的实验来看TRT node不一定能提速，但是量化确实可以，你试试FP6和INT8

**Comment by MachineJeff** [comment]
@Eloring 量化 是用什么工具呢？
@elvys-zhang 有trt的node的，但是没有提速，可能还慢了

**Comment by Eloring** [comment]
@MachineJeff 就是降低精度`precision_mode="INT8"`

**Comment by MachineJeff** [comment]
@Eloring 你可以看看我的issue，我试过 INT8，还是不行 哭了

**Comment by Eloring** [comment]
@MachineJeff 我看了你的issue，你没用他这个[repo](https://github.com/tensorflow/tensorrt/tree/master/tftrt/examples/object_detection)提供的方法，你写得那么简单应该是不行的

**Comment by MachineJeff** [comment]
@Eloring 好的 我继续 改进一下，非常感谢你的 这么长的回复，爱了爱了

**Comment by nanawe** [comment]
> @xxllp 提高的不多...
> 
> Platform	Model	Img amount	Img Size	Batch size  	mAP	Latency [load image + inference + save result]
> TF:Serving	Faster RCNN	4952 	300*300 	32 	0.7021	419204 ms
> TF:Serving +TRT	Faster RCNN	4952 	300*300 	32 	0.7019	402723 ms

这个是gpu 还是 cpu, 这个cpu和gpu都会提高么？最终结论是啥？什么情景能提速啊？

**Comment by nanawe** [comment]
> @MachineJeff 我实验发现TRT node不一定能提速，但是量化确实可以，你试试FP16和INT8

什么情况下能提速啊
